### PR TITLE
Issue #10121: Inconsistent 'TYPE_ARGUMENT' AST

### LINKED ISSUE #10121: Inconsistent `TYPE_ARGUMENT` AST
The `TYPE_ARGUMENT` AST  is not consistent when it is inside of a `TYPE_UPPER_BOUNDS` AST:

```bash


➜  src cat Test.java
import java.util.List;

public class Test<T> {
    List<String> myList;
    List<T> myOtherList;

    List<String> listMethod1() {
        return null;
    }
    List<T> listMethod2() {
        return null;
    }

    public <E extends Enum<E>> void m22() {}

}

➜  src javac Test.java

➜  src java -jar checkstyle-8.44-SNAPSHOT-all.jar -t Test.java
IMPORT -> import [1:0]
|--DOT -> . [1:16]
|   |--DOT -> . [1:11]
|   |   |--IDENT -> java [1:7]
|   |   `--IDENT -> util [1:12]
|   `--IDENT -> List [1:17]
`--SEMI -> ; [1:21]
CLASS_DEF -> CLASS_DEF [3:0]
|--MODIFIERS -> MODIFIERS [3:0]
|   `--LITERAL_PUBLIC -> public [3:0]
|--LITERAL_CLASS -> class [3:7]
|--IDENT -> Test [3:13]
|--TYPE_PARAMETERS -> TYPE_PARAMETERS [3:17]
|   |--GENERIC_START -> < [3:17]
|   |--TYPE_PARAMETER -> TYPE_PARAMETER [3:18]
|   |   `--IDENT -> T [3:18]
|   `--GENERIC_END -> > [3:19]
`--OBJBLOCK -> OBJBLOCK [3:21]
    |--LCURLY -> { [3:21]
    |--VARIABLE_DEF -> VARIABLE_DEF [4:4]
    |   |--MODIFIERS -> MODIFIERS [4:4]
    |   |--TYPE -> TYPE [4:4]
    |   |   |--IDENT -> List [4:4]
    |   |   `--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [4:8]
    |   |       |--GENERIC_START -> < [4:8]
    |   |       |--TYPE_ARGUMENT -> TYPE_ARGUMENT [4:9]  // no TYPE node
    |   |       |   `--IDENT -> String [4:9]
    |   |       `--GENERIC_END -> > [4:15]
    |   |--IDENT -> myList [4:17]
    |   `--SEMI -> ; [4:23]
    |--VARIABLE_DEF -> VARIABLE_DEF [5:4]
    |   |--MODIFIERS -> MODIFIERS [5:4]
    |   |--TYPE -> TYPE [5:4]
    |   |   |--IDENT -> List [5:4]
    |   |   `--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [5:8]
    |   |       |--GENERIC_START -> < [5:8]
    |   |       |--TYPE_ARGUMENT -> TYPE_ARGUMENT [5:9]  // no TYPE node
    |   |       |   `--IDENT -> T [5:9]
    |   |       `--GENERIC_END -> > [5:10]
    |   |--IDENT -> myOtherList [5:12]
    |   `--SEMI -> ; [5:23]
    |--METHOD_DEF -> METHOD_DEF [7:4]
    |   |--MODIFIERS -> MODIFIERS [7:4]
    |   |--TYPE -> TYPE [7:4]
    |   |   |--IDENT -> List [7:4]
    |   |   `--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [7:8]
    |   |       |--GENERIC_START -> < [7:8]
    |   |       |--TYPE_ARGUMENT -> TYPE_ARGUMENT [7:9]  // no TYPE node
    |   |       |   `--IDENT -> String [7:9]
    |   |       `--GENERIC_END -> > [7:15]
    |   |--IDENT -> listMethod1 [7:17]
    |   |--LPAREN -> ( [7:28]
    |   |--PARAMETERS -> PARAMETERS [7:29]
    |   |--RPAREN -> ) [7:29]
    |   `--SLIST -> { [7:31]
    |       |--LITERAL_RETURN -> return [8:8]
    |       |   |--EXPR -> EXPR [8:15]
    |       |   |   `--LITERAL_NULL -> null [8:15]
    |       |   `--SEMI -> ; [8:19]
    |       `--RCURLY -> } [9:4]
    |--METHOD_DEF -> METHOD_DEF [10:4]
    |   |--MODIFIERS -> MODIFIERS [10:4]
    |   |--TYPE -> TYPE [10:4]
    |   |   |--IDENT -> List [10:4]
    |   |   `--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [10:8]
    |   |       |--GENERIC_START -> < [10:8]
    |   |       |--TYPE_ARGUMENT -> TYPE_ARGUMENT [10:9]   // no TYPE node
    |   |       |   `--IDENT -> T [10:9]
    |   |       `--GENERIC_END -> > [10:10]
    |   |--IDENT -> listMethod2 [10:12]
    |   |--LPAREN -> ( [10:23]
    |   |--PARAMETERS -> PARAMETERS [10:24]
    |   |--RPAREN -> ) [10:24]
    |   `--SLIST -> { [10:26]
    |       |--LITERAL_RETURN -> return [11:8]
    |       |   |--EXPR -> EXPR [11:15]
    |       |   |   `--LITERAL_NULL -> null [11:15]
    |       |   `--SEMI -> ; [11:19]
    |       `--RCURLY -> } [12:4]
    |--METHOD_DEF -> METHOD_DEF [14:4]
    |   |--MODIFIERS -> MODIFIERS [14:4]
    |   |   `--LITERAL_PUBLIC -> public [14:4]
    |   |--TYPE_PARAMETERS -> TYPE_PARAMETERS [14:11]
    |   |   |--GENERIC_START -> < [14:11]
    |   |   |--TYPE_PARAMETER -> TYPE_PARAMETER [14:12]
    |   |   |   |--IDENT -> E [14:12]
    |   |   |   `--TYPE_UPPER_BOUNDS -> extends [14:14]
    |   |   |       |--IDENT -> Enum [14:22]
    |   |   |       `--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [14:26]
    |   |   |           |--GENERIC_START -> < [14:26]
    |   |   |           |--TYPE_ARGUMENT -> TYPE_ARGUMENT [14:27]
    |   |   |           |   `--TYPE -> TYPE [14:27]      // TYPE node is inconsistent
    |   |   |           |       `--IDENT -> E [14:27]
    |   |   |           `--GENERIC_END -> > [14:28]
    |   |   `--GENERIC_END -> > [14:29]
    |   |--TYPE -> TYPE [14:31]
    |   |   `--LITERAL_VOID -> void [14:31]
    |   |--IDENT -> m22 [14:36]
    |   |--LPAREN -> ( [14:39]
    |   |--PARAMETERS -> PARAMETERS [14:40]
    |   |--RPAREN -> ) [14:40]
    |   `--SLIST -> { [14:42]
    |       `--RCURLY -> } [14:43]
    `--RCURLY -> } [16:0


```

### PR DESCRIPTION
**Pull request by nmancus1**
Closes #10121 

### Check Regression Reports

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_checks-nonjavadoc-error/index.html

Clean 

--------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_checks-only-javadoc-error/index.html

Clean 

--------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_part1/index.html

Clean 

--------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_part2/index.html

Clean 

--------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_part3/index.html

Clean 

--------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_part4/index.html

Changes in ClassFanOutComplexityCheck, elimination of false positive:

`issue-10121`:

```bash

[BaseImmutableMultimap, BiMap, Entry, GwtCompatible, GwtIncompatible, ImmutableBiMap, ImmutableMultiset, ListMultimap, Multimap, Multiset, NavigableMap, NavigableSet, Nullable, RandomAccess, SetMultimap, SortedSetMultimap, SynchronizedBiMap, SynchronizedCollection, SynchronizedDeque, SynchronizedEntry, SynchronizedList, SynchronizedListMultimap, SynchronizedMap, SynchronizedMultimap, SynchronizedMultiset, SynchronizedNavigableMap, SynchronizedNavigableSet, SynchronizedQueue, SynchronizedRandomAccessList, SynchronizedSet, SynchronizedSetMultimap, SynchronizedSortedMap, SynchronizedSortedSet, SynchronizedSortedSetMultimap, SynchronizedTable, Table, VisibleForTesting]
[ERROR] /home/nick/Desktop/Tester/src/Test.java:65:1: Class Fan-Out Complexity is 38 (max allowed is 20). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.

```

`master`:

```bash

[BaseImmutableMultimap, BiMap, E, Entry, GwtCompatible, GwtIncompatible, ImmutableBiMap, ImmutableMultiset, ListMultimap, Multimap, Multiset, NavigableMap, NavigableSet, Nullable, RandomAccess, SetMultimap, SortedSetMultimap, SynchronizedBiMap, SynchronizedCollection, SynchronizedDeque, SynchronizedEntry, SynchronizedList, SynchronizedListMultimap, SynchronizedMap, SynchronizedMultimap, SynchronizedMultiset, SynchronizedNavigableMap, SynchronizedNavigableSet, SynchronizedQueue, SynchronizedRandomAccessList, SynchronizedSet, SynchronizedSetMultimap, SynchronizedSortedMap, SynchronizedSortedSet, SynchronizedSortedSetMultimap, SynchronizedTable, Table, VisibleForTesting]
[ERROR] /home/nick/Desktop/Tester/src/Test.java:65:1: Class Fan-Out Complexity is 38 (max allowed is 20). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.

```

The `E` type parameter should not have been counted as extra complexity. We count complexity as a class reference, a class that is imported with a direct import, or fully qualified name.

Test case added at https://github.com/checkstyle/checkstyle/blob/292188984de37e3d2f21a3a8905d2978b6704809/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityRemoveIncorrectTypeParameter.java

--------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_part5/index.html

Clean 

--------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_part6/index.html

Clean 

--------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_sevntu-check-regression_part_1/index.html

Regression in `StaticMethodCandidate`, PR opened at https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/854

--------------------------

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/diff_sevntu-check-regression_part_2/index.html

Clean

### AST Regression Reports

https://nmancus1.github.io/issue-10121_check_diff_reports_2021_06_16/antlr-diff/index.html

-------------------------

Diff Regression config: https://gist.githubusercontent.com/nmancus1/4f6ddcab47b78eb3bab2117bcee66ff2/raw/b1b24ef9a05116dd7faabd299b8b24675c8baab7/config.xml

Report label: NoWhitespaceAfter

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java [hunk 1]
```diff
@@ -460,7 +460,7 @@ private static DetailAST getPreviousNodeWithParentOfTypeAst(DetailAST ast, Detai
         // checked whether ident or lastTypeNode has preceding position
         // determining if it is java style or C style
 
-        if (ident == null || isTypeCast || ident.getLineNo() > ast.getLineNo()) {
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java:463]: We do not need this null check anymore, it was from this the inconsistent `TYPE` node that we are eliminating:

Code like:

```java

        if (info != null)
            return (TypeVariable<Class<String>>[])info.getTypeParameters();
        else
            return (TypeVariable<Class<String>>[])new TypeVariable<?>[0];
```

If we place a print statement inside this condition:

```java

        if (ident == null || isTypeCast || ident.getLineNo() > ast.getLineNo()) {
            if (ident == null && !isTypeCast) {
                System.out.println(lastTypeNode);
            }

```

We were only catching the case where the type argument child was NOT an `IDENT`:

```

>[161x45]
>[163x45]

```



From master:

```
...
    |       |   |   |       |   |--IDENT -> TypeVariable [161:20]
    |       |   |   |       |   `--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [161:32]
    |       |   |   |       |       |--GENERIC_START -> < [161:32]
    |       |   |   |       |       |--TYPE_ARGUMENT -> TYPE_ARGUMENT [161:33]
    |       |   |   |       |       |   `--TYPE -> TYPE [161:33]
    |       |   |   |       |       |       |--IDENT -> Class [161:33]
    |       |   |   |       |       |       |--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [161:38]
    |       |   |   |       |       |       |   |--GENERIC_START -> < [161:38]
    |       |   |   |       |       |       |   |--TYPE_ARGUMENT -> TYPE_ARGUMENT [161:39]
    |       |   |   |       |       |       |   |   `--TYPE -> TYPE [161:39]
    |       |   |   |       |       |       |   |       `--IDENT -> String [161:39]
    |       |   |   |       |       |       |   `--GENERIC_END -> > [161:45]
    |       |   |   |       |       |       `--ARRAY_DECLARATOR -> [ [161:47]
    |       |   |   |       |       |           `--RBRACK -> ] [161:48]
    |       |   |   |       |       `--GENERIC_END -> > [161:46]
...

```

From 10121:

```

...
    |       |   |   |       |   |--IDENT -> TypeVariable [161:20]
    |       |   |   |       |   `--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [161:32]
    |       |   |   |       |       |--GENERIC_START -> < [161:32]
    |       |   |   |       |       |--TYPE_ARGUMENT -> TYPE_ARGUMENT [161:33]
    |       |   |   |       |       |   |--IDENT -> Class [161:33]
    |       |   |   |       |       |   |--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [161:38]
    |       |   |   |       |       |   |   |--GENERIC_START -> < [161:38]
    |       |   |   |       |       |   |   |--TYPE_ARGUMENT -> TYPE_ARGUMENT [161:39]
    |       |   |   |       |       |   |   |   `--IDENT -> String [161:39]
    |       |   |   |       |       |   |   `--GENERIC_END -> > [161:45]
    |       |   |   |       |       |   `--ARRAY_DECLARATOR -> [ [161:47]
    |       |   |   |       |       |       `--RBRACK -> ] [161:48]
    |       |   |   |       |       `--GENERIC_END -> > [161:46]
 ...

```

Now we are able to get the `IDENT` from the `TYPE_ARGUMENT` ast.

**Comment by nmancus1** [comment]
Github, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterNewGenericTypeArgument.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    public static int run(String args[],PrintStream out) {
+        int i [], j [];
+        SomeInterface u [];
+        SomeInterface v[] [] = null;
+        AnotherInterface w [];
+        SomeClass x [] [];
+
+        x = (SomeClass [] []) v;
+
+        x[0] = (SomeClass []) new ImmediateSubclass [4];
```
Comment (nmancus1) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterNewGenericTypeArgument.java:64]: `null` check in NoWhitespaceAfter was not covered anymore with existing tests (all existing tests relied on extra `TYPE` ast in `TYPE_ARGUMENTS` ast), so we needed to add these to keep pitest happy.

**Comment by nmancus1** [comment]
Github, generate report

**Review by pbludov — APPROVED** [review]

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 3]
```diff
@@ -346,14 +346,14 @@ typeArguments[boolean addImagNode]
         lt:LT {#lt.setType(GENERIC_START); ;ltCounter++;}
         // (Dinesh Bolkensteyn) Added support for Java 7 diamond notation
         // (disabled ambiguous warnings since generated code seems to work)
-        (options{generateAmbigWarnings=false;}:typeArgument[addImagNode]
+        (options{generateAmbigWarnings=false;}:typeArgument[false]
```
Comment (strkkk) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:349]: it looks like argument `addImagNode` is not needed on line 342.
Same for typeArgument node (line 326)

Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:349]: I have remove the argument from `typeArguments` but the argument for `typeArgument` is still required for the calls to the other "type" productions, i.e. `classTypeSpec`, `builtInTypeSpec`, and `wildcardType`.

Comment (strkkk) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:349]: the only invocations of typeArgument are in this node and its argument is always false. So argument can be removed and usages inside typeArgument changed to false.

Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:349]: Good point. Done.

**Review by strkkk — APPROVED** [review]

**Review by pbludov — APPROVED** [review]

**Review by rnveach — APPROVED** [review]

**Review by romani — APPROVED** [review]

**Comment by romani** [comment]
Github, rebase

**Pull request merged**